File: src/omnia/data/structure/observable/writable/WritableObservableDirectedGraph.java
package omnia.data.structure.observable.writable;

import omnia.data.structure.DirectedGraph;
import omnia.data.structure.mutable.MutableDirectedGraph;
import omnia.data.structure.observable.ObservableDirectedGraph;

public interface WritableObservableDirectedGraph<E>
    extends MutableDirectedGraph<E>, ObservableDirectedGraph<E>, WritableObservableGraph<E> {

  static <E> WritableObservableDirectedGraph<E> create() {
    return new WritableObservableDirectedGraphImpl<>();
  }

  static <E> WritableObservableDirectedGraph<E> copyOf(DirectedGraph<E> original) {
    return new WritableObservableDirectedGraphImpl<>(original);
  }

  @Override
  ObservableDirectedGraph<E> toReadOnly();
}

File: src/omnia/data/structure/observable/writable/WritableObservableDirectedGraph.kt
package omnia.data.structure.observable.writable

import omnia.data.structure.DirectedGraph
import omnia.data.structure.mutable.MutableDirectedGraph
import omnia.data.structure.observable.ObservableDirectedGraph

interface WritableObservableDirectedGraph<E : Any>
  : MutableDirectedGraph<E>, ObservableDirectedGraph<E>, WritableObservableGraph<E> {

  override fun toReadOnly(): ObservableDirectedGraph<E>

  companion object {

    @JvmStatic
    fun <E : Any> create(): WritableObservableDirectedGraph<E> {
      return WritableObservableDirectedGraphImpl()
    }

    fun <E : Any> copyOf(original: DirectedGraph<E>): WritableObservableDirectedGraph<E> {
      return WritableObservableDirectedGraphImpl(original)
    }
  }
}